### PR TITLE
[BEAM-9258] Integrate Google Cloud Data loss prevention functionality for Python SDK

### PR DESCRIPTION
**Pull request by mwalenia**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mwalenia** [comment]
R: @aaltay Can you take a look at this? Thanks!

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    self.project = project
+
+  def start_bundle(self):
+    if self.client is None:
+      self.client = dlp_v2.DlpServiceClient()
+
+  def process(self, element, **kwargs):
+    params = {
+        'timeout': self.timeout,
+        'retry': retry.with_exponential_backoff(
```
Comment (aaltay) [inline sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: retry argument is defined as 

```
retry (Optional[google.api_core.retry.Retry]) – A retry object used to retry requests. If None is specified, requests will be retried using a default configuration.
```

(https://googleapis.dev/python/dlp/latest/gapic/v2/api.html)

I do not believe retry.with_exponential_backoff returns a object of type google.api_core.retry.Retry

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    self.timeout = timeout
+    self.client = client
+    self.project = project
+
+  def start_bundle(self):
+    if self.client is None:
+      self.client = dlp_v2.DlpServiceClient()
+
+  def process(self, element, **kwargs):
+    params = {
```
Comment (aaltay) [inline sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: params and params.update, could it move to start_bundle?

File: sdks/python/setup.py [hunk 3]
```diff
@@ -203,6 +203,7 @@ def get_version():
     'google-cloud-bigquery>=1.6.0,<1.18.0',
     'google-cloud-core>=0.28.1,<2',
     'google-cloud-bigtable>=0.31.1,<1.1.0',
+    'google-cloud-dlp >=0.12.0,<=0.13.0',
```
Comment (aaltay) [inline sdks/python/setup.py]: Version after 0.13 will not support py2. (notice at the top: https://googleapis.dev/python/dlp/latest/gapic/v2/api.html)

I wonder if we need to add a comment note here for the person that will upgrade version ranges next?

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py (hunk 1 above)
Comment (mwalenia) [reply sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: You're right. Let's stick with the default here.

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  """Inspects input text for sensitive information.
+  the ``PTransform`` returns a ``PCollection`` of
+  ``List[google.cloud.dlp_v2.proto.dlp_pb2.Finding]``
+  Example usage::
+      pipeline | InspectForDetails(project='example-gcp-project',
+                inspection_config={'info_types': [{'name': 'EMAIL_ADDRESS'}]})
+  """
+  def __init__(
+      self,
+      inspection_template_name=None,
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: Yeah, I'll rearrange it

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    functionality.
+"""
+
+from __future__ import absolute_import
+
+import logging
+
+from google.cloud import dlp_v2
+
+import apache_beam as beam
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: Ok!

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py (hunk 2 above)
Comment (mwalenia) [reply sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: I moved it to `setup()`. Will that be OK?

**Comment by mwalenia** [comment]
@aaltay Thanks for the review, I fixed the code so that `--project` pipeline option is used for project name.

I also changed the way arguments were handled to reflect their meaning in underlying DLP service - template name and config object are allowed simultaneously, but config overrides the template.

**Comment by mwalenia** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by mwalenia** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/ml/gcp/cloud_dlp.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@experimental()
+class MaskDetectedDetails(PTransform):
+  """Scrubs sensitive information detected in text.
+  The ``PTransform`` returns a ``PCollection`` of ``str``
+  Example usage::
+    pipeline | MaskDetectedDetails(project='example-gcp-project',
```
Comment (aaltay) [inline sdks/python/apache_beam/ml/gcp/cloud_dlp.py]: You may need to update this example for the project argument.

**Comment by aaltay** [comment]
LGTM.

Related to `project`. Do you think we should leave it and default it to the project from the flags. The reason is, people sometimes want to use different projects for other resources. (e.g. run dataflow job on project X, make api calls to project Y.) Making it have a default but configurable would help users.

**Comment by mwalenia** [comment]
Run Python PreCommit

**Comment by mwalenia** [comment]
@aaltay Thanks for your suggestions, I added the `project` param back to the classes. WDYT now?

**Pull request merged**